utils/state.py

```diff
@@ -48,6 +48,14 @@ def exists(self, key):
         except ClientError:
             return False
 
+    def ls(self):
```

Comment:
`__iter__`? so one can do `for key in state:`

Reply:
the state object represents a remote state. i'd like this class to emphasize that the state is remote in every way possible.

`state.ls()` to me shows that some actions will be carried out here. `for key in state` feels so in-memory. maybe it's just me.

Reply:
I see. Never thought of `__iter__` as an interface for in-memory content. Actualy I see it more as the opposite... it typically creates a generator so the items are retrieved one by one, from wherever they are. Eg `f = open('foo.txt')`, the `for line in f` will give you one line at a time from the file. Anyway, aways fun to think about interfaces :)

Reply:
__iter__ itself, yes.
but when using state without looking in the class, `for key in state`?

Reply:
yep... I see just a loop over the state elements, wherever they are coming from. But I get it, we see it differently. No worries. Too bad that those interfaces got worn out by python's own data structures, now it's hard to think of them differently.
To be clear, I'm ok with the `ls()`.